### PR TITLE
feat: teritori contracts

### PR DESCRIPTION
**Pull request by n0izn0iz**
This is a snapshot of the contracts developed by the Teritori team

See the [overview](https://github.com/gnolang/gno/blob/3b030130c2639a4ea3570c2074b8ff447807ccdb/examples/gno.land/r/demo/teritori/README.md) for more information

<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

File: examples/gno.land/r/demo/teritori/modboards/gno.mod [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module gno.land/r/demo/teritori/modboards
+
+require (
+        "gno.land/p/demo/avl" v0.0.0-latest
+        "gno.land/r/demo/users" v0.0.0-latest
+        "gno.land/p/demo/teritori/dao_interfaces" v0.0.0-latest
+	"gno.land/p/demo/teritori/ujson" v0.0.0-latest
```
Comment (zivkovicmilos) [inline examples/gno.land/r/demo/teritori/modboards/gno.mod:7]: Am I crazy or is this not formatted correctly?

File: examples/gno.land/r/demo/teritori/justicedao/justicedao.gno [hunk 2]
```diff
@@ -0,0 +1,742 @@
...
+	caller := std.GetOrigCaller()
+
+	// if sender is not a dao member, revert
+	isCallerDaoMember := IsDAOMember(caller)
+	if !isCallerDaoMember {
+		panic("caller is not a gnodao member")
+	}
+
+	// TODO: send coins from caller to DAO
+	// TODO: verify received amount
```
Comment (zivkovicmilos) [inline examples/gno.land/r/demo/teritori/justicedao/justicedao.gno:368]: Leftover?

File: examples/gno.land/r/demo/teritori/justicedao/justicedao.gno [hunk 3]
```diff
@@ -0,0 +1,742 @@
...
+	"uri": "%s",
+	"metadata": "%s",
+	"funds" %d,
+	"totalVotingPower": %d,
+	"votingPeriod": %d,
+	"voteQuorum": %d,
+	"threshold": %d,
+	"vetoThreshold": %d,
+	"numJusticeDAO": %d
+}`, daoEncode.uri, daoEncode.metadata, daoEncode.funds, daoEncode.totalVotingPower, daoEncode.votingPeriod, daoEncode.voteQuorum, daoEncode.threshold, daoEncode.vetoThreshold, daoEncode.numJusticeDAO)
```
Comment (zivkovicmilos) [inline examples/gno.land/r/demo/teritori/justicedao/justicedao.gno:474]: Nitpick: can we format this into multiple lines?

File: examples/gno.land/r/demo/teritori/groups/public.gno [hunk 4]
```diff
@@ -0,0 +1,126 @@
+package groups
+
+import (
+	"std"
+
+	"gno.land/p/demo/teritori/ujson"
+	// "gno.land/r/demo/users"
```
Comment (zivkovicmilos) [inline examples/gno.land/r/demo/teritori/groups/public.gno:7]: Leftover comment?

File: examples/gno.land/r/demo/teritori/groups/public.gno [hunk 5]
```diff
@@ -0,0 +1,126 @@
+package groups
```
Comment (zivkovicmilos) [inline examples/gno.land/r/demo/teritori/groups/public.gno:1]: Please check the leftover comments in this file 🙏

File: examples/gno.land/r/demo/teritori/groups/README.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
+### - test package
```
Comment (zivkovicmilos) [inline examples/gno.land/r/demo/teritori/groups/README.md:1]: Same comment as before -- what do you think about converting this into a script?

File: examples/gno.land/r/demo/teritori/escrow/escrow_test.gno [hunk 7]
```diff
@@ -0,0 +1 @@
+package escrow
```
Comment (zivkovicmilos) [inline examples/gno.land/r/demo/teritori/escrow/escrow_test.gno:1]: Why are tests intentionally missing?

File: examples/gno.land/r/demo/teritori/dao_registry/dao_registry.gno [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+import (
+	"encoding/binary"
+	"std"
+	"time"
+
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/teritori/ujson"
+)
+
+// TODO: rewrite
```
Comment (zivkovicmilos) [inline examples/gno.land/r/demo/teritori/dao_registry/dao_registry.gno:12]: What does this mean?

File: examples/gno.land/p/demo/teritori/ujson/parse.gno [hunk 9]
```diff
@@ -0,0 +1,594 @@
+package ujson
```
Comment (zivkovicmilos) [inline examples/gno.land/p/demo/teritori/ujson/parse.gno:1]: /summon @thehowl for thoughts on this package

File: examples/gno.land/p/demo/teritori/markdown_utils/markdown_utils.gno [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+			// add an indentation level to the title
+			lines[i] = "#" + line
+		}
+	}
+
+	// join the lines back into a string
+	return strings.Join(lines, "\n")
+}
+
+// thanks copilot this is perfect xD
```
Comment (zivkovicmilos) [inline examples/gno.land/p/demo/teritori/markdown_utils/markdown_utils.gno:25]: Leftover comments? :)

File: examples/gno.land/p/demo/teritori/havl/havl.gno [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+	"gno.land/p/demo/avl"
+)
+
+type Tree struct {
+	root          avl.Tree // height -> *avl.Tree
+	initialHeight int64
+}
+
+// FIXME: don't cast height to int
```
Comment (zivkovicmilos) [inline examples/gno.land/p/demo/teritori/havl/havl.gno:16]: What does this mean?

File: examples/gno.land/p/demo/teritori/flags_index/flags_index.gno [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+type FlagID string
+
+type FlagCount struct {
+	FlagID FlagID
+	Count  uint64
+}
+
+type FlagsIndex struct {
+	flagsCounts      []*FlagCount // sorted by count descending; TODO: optimize using big brain datastructure
```
Comment (zivkovicmilos) [inline examples/gno.land/p/demo/teritori/flags_index/flags_index.gno:17]: Probably one of the best TODOs we have in the codebase; please leave it

File: examples/gno.land/p/demo/teritori/dao_proposal_single/update_settings.gno [hunk 13]
```diff
@@ -0,0 +1,78 @@
+package dao_proposal_single
+
+import (
+	"strings"
+
+	"gno.land/p/demo/teritori/dao_interfaces"
+	"gno.land/p/demo/teritori/ujson"
+)
+
+// TODO: convert to json
```
Comment (zivkovicmilos) [inline examples/gno.land/p/demo/teritori/dao_proposal_single/update_settings.gno:10]: Is this TODO valid anymore?

File: examples/gno.land/p/demo/teritori/dao_proposal_single/threshold.gno [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+type Threshold interface {
+	Clone() Threshold
+	ToJSON() string
+}
+
+func ThresholdFromJSON(ast *ujson.JSONASTNode) Threshold {
+	ac := ThresholdAbsoluteCount(0)
+	return ast.ParseUnion([]*ujson.ParseKV{
+		// TODO: {Key: "absolutePercentage"},
```
Comment (zivkovicmilos) [inline examples/gno.land/p/demo/teritori/dao_proposal_single/threshold.gno:73]: Leftover?

File: examples/gno.land/p/demo/teritori/dao_core/dao_core.gno [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+// mutations
+
+func (d *daoCore) UpdateVotingModule(newVotingModule dao_interfaces.IVotingModule) {
+	if std.CurrentRealm().Addr() != d.realm.Addr() { // not sure this check necessary since the ownership system should protect against mutation from other realms
+		panic(ErrUnauthorized)
+	}
+
+	// FIXME: check da0-da0 implem
```
Comment (zivkovicmilos) [inline examples/gno.land/p/demo/teritori/dao_core/dao_core.gno:79]: Leftover comment?

**Review by zivkovicmilos — APPROVED** [review]
Thank you for completing these milestones 🙏 

I have left some comments, mostly just questions and pointers to leftover code segments.

Calling @gfanton and @ajnavarro to also take a look through the code.

What you can't see from this codebase is how this integrates with Teritori, @n0izn0iz are there any Looms / recordings of these contracts in action on the Teritori app? We did showcases live on calls, but I would love the rest of the team to see them as well.

Amazing work 👏

File: examples/gno.land/p/demo/teritori/utf16/utf16.gno
Comment (gfanton) [inline examples/gno.land/p/demo/teritori/utf16/utf16.gno:1]: I think we should move this to gnovm/stdlibs/unicode/utf16, along with theirs respective [tests](https://cs.opensource.google/go/go/+/master:src/unicode/utf16/).
And update [compatibility file](https://github.com/gnolang/gno/blob/fa8eb7753dc5396144dd4166123234af3a3a4da1/gnovm/docs/go-gno-compatibility.md?plain=1#L340C1-L340C1) on the way

**Comment by n0izn0iz** [comment]
Closing to atomicize

File: examples/gno.land/r/demo/teritori/tori/tori.gno [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+var (
+	tori     *grc20.AdminToken
+	userTori grc20.IGRC20
+	admin    std.Address = std.DerivePkgAddr("gno.land/r/demo/teritori/dao_realm") // TODO: helper to change admin
+)
+
+func init() {
+	tori = grc20.NewAdminToken("Tori", "TORI", 6)
+	userTori = tori.GRC20()
+	// grc20_registry.Register(userTori) // found another bug, calling this sets some grc20 internal object's pkgid to the registry realm id
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply examples/gno.land/r/demo/teritori/tori/tori.gno:22]: https://github.com/gnolang/gno/issues/974

File: examples/gno.land/r/demo/teritori/tori/tori.gno [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+	caller := std.PrevRealm().Addr()
+	tori.TransferFrom(caller, from.Resolve(), to.Resolve(), amount)
+}
+
+// faucet.
+
+func Faucet() {
+	// FIXME: add limits?
+	// FIXME: add payment in gnot?
+	caller := std.PrevRealm().Addr()
```
[the remark this answers is not in the document]
Comment (n0izn0iz) [reply examples/gno.land/r/demo/teritori/tori/tori.gno:72]: yep from the copy of grc20

File: examples/gno.land/p/demo/teritori/dao_proposal_single/update_settings.gno (hunk 13 above)
Comment (n0izn0iz) [reply examples/gno.land/p/demo/teritori/dao_proposal_single/update_settings.gno:10]: nope

**Pull request closed without merging**